### PR TITLE
Only use public `deep_copy` inside containers

### PR DESCRIPTION
**Pull request by pzehner**
This PR is a refactor which is part of a greater effort to simplify the use of `Impl::DeepCopy`.

It replaces the `Impl::DeepCopy` used in containers by the public `deep_copy`.

### CODE REVIEW

File: containers/src/Kokkos_Bitset.hpp [hunk 1]
```diff
@@ -135,9 +135,13 @@ class Bitset {
 
     if (m_last_block_mask) {
       // clear the unused bits in the last block
-      Kokkos::Impl::DeepCopy<typename Device::memory_space, Kokkos::HostSpace>(
-          m_blocks.data() + (m_blocks.extent(0) - 1u), &m_last_block_mask,
-          sizeof(unsigned));
+      Kokkos::parallel_for(
+          "clear the unused bits in the last block",
```
Comment (Rombur) [inline containers/src/Kokkos_Bitset.hpp:139]: The label should be prefixed with `Kokkos::`

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 2]
```diff
@@ -901,33 +924,34 @@ class UnorderedMap {
   bool modified() const { return get_flag(modified_idx); }
 
   void set_flag(int flag) const {
-    using raw_deep_copy =
-        Kokkos::Impl::DeepCopy<typename device_type::memory_space,
-                               Kokkos::HostSpace>;
-    const int true_ = true;
-    raw_deep_copy(m_scalars.data() + flag, &true_, sizeof(int));
+    Kokkos::parallel_for(
+        "set flag",
```
Comment (Rombur) [inline containers/src/Kokkos_UnorderedMap.hpp:928]: Here and below prefix the label  with `Kokkos::`

File: containers/src/Kokkos_Bitset.hpp (hunk 1 above)
Comment (pzehner) [reply containers/src/Kokkos_Bitset.hpp]: Done.

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 2 above)
Comment (pzehner) [reply containers/src/Kokkos_UnorderedMap.hpp]: Done.

**Comment by pzehner** [comment]
I wonder if the fences are still necessary.

File: containers/src/Kokkos_Bitset.hpp [hunk 3]
```diff
@@ -402,10 +406,7 @@ void deep_copy(Bitset<DstDevice>& dst, Bitset<SrcDevice> const& src) {
   }
 
   Kokkos::fence("Bitset::deep_copy: fence before copy operation");
-  Kokkos::Impl::DeepCopy<typename DstDevice::memory_space,
-                         typename SrcDevice::memory_space>(
-      dst.m_blocks.data(), src.m_blocks.data(),
-      sizeof(unsigned) * src.m_blocks.extent(0));
+  Kokkos::deep_copy(dst.m_blocks, src.m_blocks);
   Kokkos::fence("Bitset::deep_copy: fence after copy operation");
```
Comment (masterleinad) [inline containers/src/Kokkos_Bitset.hpp:410]: You don't eed the fences here anymore.

File: containers/src/Kokkos_Bitset.hpp [hunk 4]
```diff
@@ -417,10 +418,7 @@ void deep_copy(Bitset<DstDevice>& dst, ConstBitset<SrcDevice> const& src) {
...
```
Comment (masterleinad) [inline containers/src/Kokkos_Bitset.hpp:422]: These can go as well.

File: containers/src/Kokkos_Bitset.hpp [hunk 5]
```diff
@@ -432,10 +430,7 @@ void deep_copy(ConstBitset<DstDevice>& dst, ConstBitset<SrcDevice> const& src) {
...
```
Comment (masterleinad) [inline containers/src/Kokkos_Bitset.hpp:434]: ... and here

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 6]
```diff
@@ -874,22 +906,13 @@ class UnorderedMap {
...
-        raw_deep_copy(m_values.data(), src.m_values.data(),
-                      sizeof(impl_value_type) * src.m_values.extent(0));
+        Kokkos::deep_copy(m_values, src.m_values);
       }
-      raw_deep_copy(m_scalars.data(), src.m_scalars.data(),
-                    sizeof(int) * num_scalars);
+      Kokkos::deep_copy(m_scalars, src.m_scalars);
 
       Kokkos::fence(
           "Kokkos::UnorderedMap::deep_copy_view: fence after copy to dst.");
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:918]: This fence can also be removed.

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 7]
```diff
@@ -901,33 +924,34 @@ class UnorderedMap {
...
-        Kokkos::Impl::DeepCopy<Kokkos::HostSpace,
-                               typename device_type::memory_space>;
-    int result = false;
-    raw_deep_copy(&result, m_scalars.data() + flag, sizeof(int));
+    bool result = false;
+    Kokkos::parallel_reduce(
+        "Kokkos::get flag",
+        Kokkos::RangePolicy(typename device_type::execution_space{}, flag,
+                            flag + 1),
+        Impl::FlagGetFunctor(m_scalars), result);
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:954]: Why can't this be a lambda?

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 8]
```diff
@@ -901,33 +924,34 @@ class UnorderedMap {
...
-    using raw_deep_copy =
-        Kokkos::Impl::DeepCopy<typename device_type::memory_space,
-                               Kokkos::HostSpace>;
-    const int false_ = false;
-    raw_deep_copy(m_scalars.data() + flag, &false_, sizeof(int));
+    Kokkos::parallel_for(
+        "Kokkos::reset flag",
+        Kokkos::RangePolicy(typename device_type::execution_space{}, flag,
+                            flag + 1),
+        Impl::FlagSetFunctor(m_scalars, false));
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:942]: Why can't this be a lambda?

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 9]
```diff
@@ -901,33 +924,34 @@ class UnorderedMap {
...
-    using raw_deep_copy =
-        Kokkos::Impl::DeepCopy<typename device_type::memory_space,
-                               Kokkos::HostSpace>;
-    const int true_ = true;
-    raw_deep_copy(m_scalars.data() + flag, &true_, sizeof(int));
+    Kokkos::parallel_for(
+        "Kokkos::set flag",
+        Kokkos::RangePolicy(typename device_type::execution_space{}, flag,
+                            flag + 1),
+        Impl::FlagSetFunctor(m_scalars, true));
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:931]: Why can't this be a lambda?

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 10]
```diff
@@ -148,6 +148,38 @@ struct UnorderedMapInsertOpTypes {
...
+struct FlagGetFunctor {
+  View m_view;
+
+  explicit FlagGetFunctor(View const &view) : m_view(view) {}
+
+  KOKKOS_FUNCTION
+  void operator()(unsigned const i, bool &value_inner) const {
+    value_inner = static_cast<bool>(m_view(i));
+  }
+};
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:179]: These are so short that it hsould be fine to inline them.

File: containers/src/Kokkos_DynamicView.hpp [hunk 11]
```diff
@@ -195,11 +195,13 @@ struct ChunkedArrayManager {
...
-          exec_space, other.m_chunks, m_chunks,
-          sizeof(pointer_type) * (m_chunk_max + 2));
-    }
+    // use of ad-hoc unmanaged views
+    Kokkos::deep_copy(
+        exec_space,
+        Kokkos::View<value_type*, OtherMemorySpace>(
+            reinterpret_cast<value_type*>(other.m_chunks), m_chunk_max + 2),
+        Kokkos::View<value_type*, MemorySpace>(
+            reinterpret_cast<value_type*>(m_chunks), m_chunk_max + 2));
```
Comment (masterleinad) [inline containers/src/Kokkos_DynamicView.hpp:204]: Why is the `interpret_cast` the right thing to do here?

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 7 above)
Comment (pzehner) [reply containers/src/Kokkos_UnorderedMap.hpp:954]: Because when doing so, NVCC complains with:

```
error: The enclosing parent function ("get_flag") for an extended __host__ __device__ lambda cannot have private or protected access within its class
```

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 8 above)
Comment (pzehner) [reply containers/src/Kokkos_UnorderedMap.hpp:942]: Same as above.

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 9 above)
Comment (pzehner) [reply containers/src/Kokkos_UnorderedMap.hpp:931]: Same as above.

File: containers/src/Kokkos_Bitset.hpp (hunk 3 above)
Comment (pzehner) [reply containers/src/Kokkos_Bitset.hpp:410]: Done.

File: containers/src/Kokkos_Bitset.hpp (hunk 4 above)
Comment (pzehner) [reply containers/src/Kokkos_Bitset.hpp:422]: Done.

File: containers/src/Kokkos_Bitset.hpp (hunk 5 above)
Comment (pzehner) [reply containers/src/Kokkos_Bitset.hpp:434]: Done.

File: containers/src/Kokkos_DynamicView.hpp (hunk 11 above)
Comment (pzehner) [reply containers/src/Kokkos_DynamicView.hpp:204]: `m_chunks` is a double pointer that is treated as `void *` by `Dynk::DeepCopy`. `Kokkos::View`, even unmanaged, requires to set the appropriate rank and dimensions, the later being unknown. So I use `reinterpret_cast` to force treating `m_chunks` as 1-rank data. As the view is temporary and only used for the call of `deep_copy`, I think it is OK to do so.

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 10 above)
Comment (pzehner) [reply containers/src/Kokkos_UnorderedMap.hpp:179]: Do you mean using `inline`?

**Review by Rombur — APPROVED** [review]

File: containers/src/Kokkos_DynamicView.hpp (hunk 11 above)
Comment (masterleinad) [reply containers/src/Kokkos_DynamicView.hpp:204]: If I'm not mistaken, the original call did something like
```C++
memcpy(reinterpret_cast<pointer_type*>(&other.m_chunks), reinterpret_cast<pointer_type*>(&m_chunks), sizeof(pointer_type) * (m_chunk_max + 2));
```
(where the reinterpret_cast is irrelevant of course)
while the new implementation does
```C++
memcpy(reinterpret_cast<value_type*>(&other.m_chunks), reinterpret_cast<value_type*>(&m_chunks), sizeof(value_type) * (m_chunk_max + 2));
```
and these would only do the same if `sizeof(value_type) == `sizeof(pointer_type)`. In other words, the Views data type should be `pointer_type` (which is problematic since int** would be treated as a two-dimensional View) or something with the same size like `uintptr_t`.

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 7 above)
Comment (pzehner) [reply containers/src/Kokkos_UnorderedMap.hpp]: Eventually replaced it by a call to `deep_copy` + a scalar subview.

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 9 above)
Comment (pzehner) [reply containers/src/Kokkos_UnorderedMap.hpp]: Eventually replaced it by a call to `deep_copy` + a scalar subview.

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 8 above)
Comment (pzehner) [reply containers/src/Kokkos_UnorderedMap.hpp]: Eventually replaced it by a call to `deep_copy` + a scalar subview.

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 10 above)
Comment (pzehner) [reply containers/src/Kokkos_UnorderedMap.hpp]: Eventually replaced it by a call to `deep_copy` + a scalar subview.

File: containers/src/Kokkos_DynamicView.hpp (hunk 11 above)
Comment (pzehner) [reply containers/src/Kokkos_DynamicView.hpp:204]: To clarify things out, `pointer_type` is the same as `value_type*`, and `m_chunks` is a `pointer_type*`.

I see what you mean, so my PR is very likely copying a wrong amount of data (since `sizeof(value_type*) == sizeof(pointer_type)`), then?

Comment (masterleinad) [reply containers/src/Kokkos_DynamicView.hpp:204]: > I see what you mean, so my PR is very likely copying a wrong amount of data (since `sizeof(value_type*) == sizeof(pointer_type)`), then?

That's what it looks like to me.

Comment (pzehner) [reply containers/src/Kokkos_DynamicView.hpp:204]: Do you have any idea on how to fix that?

Comment (masterleinad) [reply containers/src/Kokkos_DynamicView.hpp:204]: Use `uintptr_t` for the `View`'s data type (and reinterpret_cast to `uintptr_t*` instead of `value_type*`).

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 12]
```diff
@@ -874,25 +874,13 @@ class UnorderedMap {
...
-        raw_deep_copy(m_values.data(), src.m_values.data(),
-                      sizeof(impl_value_type) * src.m_values.extent(0));
+        Kokkos::deep_copy(m_values, src.m_values);
       }
-      raw_deep_copy(m_scalars.data(), src.m_scalars.data(),
-                    sizeof(int) * num_scalars);
-
-      Kokkos::fence(
-          "Kokkos::UnorderedMap::deep_copy_view: fence after copy to dst.");
+      Kokkos::deep_copy(m_scalars, src.m_scalars);
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:883]: Note that we now have up to 12 global fences where 2 would be sufficient.

File: containers/src/Kokkos_DynamicView.hpp (hunk 11 above)
Comment (pzehner) [reply containers/src/Kokkos_DynamicView.hpp:204]: There is no guarantee that `value_type` is an `int`, since it's template parameter. In the tests I see, it's a `double`.

Comment (masterleinad) [reply containers/src/Kokkos_DynamicView.hpp:204]: `uintptr_t` does have the same size as a pointer. It doesn't matter what the type pointed to is. Also, round-trip conversion between a pointer and `uintptr_t` is fine and that's what we are effectively doing here. The `memcpy` never tries to interpret values stored in `uintptr_t`.

Comment (pzehner) [reply containers/src/Kokkos_DynamicView.hpp]: Ok, I didn't know that. Seems like it works.

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 12 above)
Comment (pzehner) [reply containers/src/Kokkos_UnorderedMap.hpp:883]: That's a lot indeed. Is it possible to call `deep_copy` without fencing?

Comment (masterleinad) [reply containers/src/Kokkos_UnorderedMap.hpp:883]: You could at least use a three-argument deep_copy for all those calls except the first and the last.

Comment (masterleinad) [reply containers/src/Kokkos_UnorderedMap.hpp:883]: > That's a lot indeed. Is it possible to call deep_copy without fencing?

That'd what the three-argument `deep_copy` is for.

File: containers/src/Kokkos_DynamicView.hpp (hunk 11 above)
Comment (pzehner) [reply containers/src/Kokkos_DynamicView.hpp]: Ew, the code segfaults now...

Comment (pzehner) [reply containers/src/Kokkos_DynamicView.hpp]: Fixed.

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 12 above)
Comment (pzehner) [reply containers/src/Kokkos_UnorderedMap.hpp]: Ok, I think I found a way.

**Review by masterleinad — APPROVED** [review]
Fine with me.

**Review by masterleinad — APPROVED** [review]
Still looks good to me.

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 13]
```diff
@@ -874,24 +874,18 @@ class UnorderedMap {
...
-        raw_deep_copy(m_values.data(), src.m_values.data(),
-                      sizeof(impl_value_type) * src.m_values.extent(0));
+        Kokkos::deep_copy(exec_space, m_values, src.m_values);
       }
-      raw_deep_copy(m_scalars.data(), src.m_scalars.data(),
-                    sizeof(int) * num_scalars);
+      Kokkos::deep_copy(exec_space, m_scalars, src.m_scalars);
 
-      Kokkos::fence(
+      exec_space.fence(
```
Comment (Rombur) [inline containers/src/Kokkos_UnorderedMap.hpp:888]: Why did you change to `exec_space`? Currently we synchronize everything (`Kokkos::fence()` calls `static_fence()` which ends up calling `cudaDeviceSynchronize`). With this change,  if you use `--default-stream per-thread` with the CUDA backend, you will only fence the default streams not the others, right?

[the remark this answers is not in the document]
Comment (pzehner) [reply containers/src/Kokkos_UnorderedMap.hpp:888]: I don't have a strong opinion about this

Comment (Rombur) [reply containers/src/Kokkos_UnorderedMap.hpp:888]: `deep_copy_view` does not take an execution space, so I would expect to have a global fence. Unless there is a good reason, I think we should revert this change.

Comment (pzehner) [reply containers/src/Kokkos_UnorderedMap.hpp:888]: Ok, I'll revert this.

**Review by Rombur — APPROVED** [review]

File: containers/src/Kokkos_Bitset.hpp [hunk 14]
```diff
@@ -400,13 +400,7 @@ void deep_copy(Bitset<DstDevice>& dst, Bitset<SrcDevice> const& src) {
     Kokkos::Impl::throw_runtime_exception(
         "Error: Cannot deep_copy bitsets of different sizes!");
   }
```
Comment (crtrott) [inline containers/src/Kokkos_Bitset.hpp:402]: Should we remove these throws here? We have checks inside Kokkos::deep_copy for matching size. Though having it here would keep the more specific message.

**Review by crtrott — APPROVED** [review]
Approving but want to give other folks a chance to weigh in on the question of the throw in Bitset.

File: containers/src/Kokkos_Bitset.hpp (hunk 14 above)
Comment (dalg24) [reply containers/src/Kokkos_Bitset.hpp:402]: Not feeling too strongly about it.  I would merge the PR as is.

Comment (crtrott) [reply containers/src/Kokkos_Bitset.hpp:402]: it would have been in scope since the check clearly accompanies the impl deep copy call same as the fences
